### ISSUE TITLE
[FEATURE] Image installation mode in the harvester installer (install via image)

### ISSUE THREAD

**Issue by ibrokethecloud** [harvester/harvester#2198]
**Is your feature request related to a problem? Please describe.**

New Installation mode for harvester to allow configuration to better enable cloud / edge use cases

**Describe the solution you'd like**

Currently when a new harvester node is launched it needs to be the first node in the cluster or join an existing cluster. 
This is very useful when we already know the details of the environment where harvester is being installed. 
However for use cases like metal cloud providers and edge, we need an additional mode, say, `install`, which will install and load the OS content onto the node. 

As part of this install the rancherd and networking configuration will not be applied.

When the user boots this "configured" node, the console will wait for the end user to configure the following:
* create/join option for rancherd
* join details if this node needs to join an existing cluster
* management network interface details.

Once this information is made available, the installer will apply the endpoint info and bootstrap rancherd + rke2

**Describe alternatives you've considered**

PXE boot using Tinkerbell boots, which works fine, however this may not be eligible in a cloud based or edge environment.
Harvester CAPI provider, which will again need a central k8s cluster with connectivity, and not to mention eventually PXE boot nodes.

**Additional context**

**Comment by abonillabeeche** [comment]
This sounds great and would be useful for appliances. We should only collect the installation disk on installation and on first boot we should ask for all the other information that we currently ask for, including the data/Longhorn disk.

**Comment by janeczku** [comment]
I'd like to see a variant of this, where during the "unconfigured" installation mode i can specify a URL which will then be fetched on the first boot and contain the join or create config.

Additionally, that URL parameter may support variable expansion, so that it can generate and fetch hardware specific URLs like: `https://config-db/harvester-node-config?mac=00:00:5e:00:53:af`

**Comment by oivindoh** [comment]
This would be quite useful to us.

**Comment by ibrokethecloud** [comment]
replaced by: https://github.com/harvester/harvester-installer/pull/457

**Comment by TachunLin** [comment]
### Test Plan 

#### Positive - Installation Mode 

Plan to validate on v1.2.0-rc3

- [x] Ensure can create image node correctly
- [x] Ensure can create Harvester in create mode using the image node 
- [x] Ensure can join Harvester in create mode using the image node 
- [x] Ensure the follow up configuration of network, dns, cluster token, password, ntp server, proxy, ssh keys, Harvester config

- [x] Validate the functionality of image node for create and join mode afterward   

  - No breaks for create and join mode 

- [x] Create / join mode cross test

  |                     | Image node for Create | Existing create mode |
  |---------------------|:---------------------:|:--------------------:|
  | Image node for Join |      PASS                 |       PASS               |
  | Existing join mode  |        PASS               |         PASS             |


#### Negative - Installation Mode 
- [x] Reboot after installed binary mode 
- [x] Shutdown / Boot after installed binary mode and test the join mode

#### Positive - Streaming mode 
Plan to validate on v1.2.0-rc4  

Test plan require further discussion.  

* Refer to: https://github.com/harvester/harvester-installer/pull/457#issuecomment-1619385668

 |                     | Streaming Create | Streaming Join |
  |---------------------|:---------------------:|:--------------------:|
  | Import storage |         PASS              |     PASS                 |
  | Manual boot option  |      PASS                 |     PASS                 |

- [x] Steaming install via manual specify boot option 
    - [x] Ensure can use the prepared raw file to install Harvester in create mode (Streaming) 
    - [x] Ensure can use the prepared raw file to install Harvester in join mode (Streaming) 

- [x] Steaming mode install via import storage 
    - [x] Ensure can use the prepared raw file to install Harvester in create mode (Streaming) 
    - [x] Ensure can use the prepared raw file to install Harvester in join mode (Streaming)

**Comment by TachunLin** [comment]
During the negative testing, found an potential issue that when we `directly shutdown` an harvester installed via binary mode. 

After we start the installed Harvester image node again. it failed to launch into the create/join mode selection console
  ![image](https://github.com/harvester/harvester/assets/29251855/efae6911-3095-4e8a-b2bf-1e2b8f96face)

Instead if we `reboot` the vm after installed via binary mode, it can proceed to the installation console 

Not sue whether it could impact the usage of the install via image mode. 
Perhaps it worth investigate more in details, thus suggest move to the **implement** stage 

### Test Information
* Test Environment: two nodes harvester on local kvm machine
* Harvester version: v1.2.0-rc3

### Verify Steps
1. Create the first vm and install a Harvester via binary mode 
2. Setup the image to create a single node Harvester (create mode)
3. Create the second vm and install the second Harvester via binary mode (used for join mode)
4. After installed via image complete and enter into create/join mode selection console 
5. Directly shutdown the vm 
6. Start the vm again

**Comment by TachunLin** [comment]
Update the streaming mode test plan and use the predefined v1.2-head image file on verification.

**Comment by TachunLin** [comment]
Verified **fixed** on `v1.2.0-rc4` and `v1.2-37aeb5ff`. Close this issue.

### Result
According to test plan in https://github.com/harvester/harvester/issues/2198#issuecomment-1637703769

We verified can work on the following types of image installation mode 

1. <details><summary>ISO installation mode $~~$ $\color{green}{\textsf{PASS}}$</summary>
   
    Can use ISO installation to create and join Harvester from the installation mode of the following scenarios 
    
     |                     | Image node for Create | Existing create mode |
    |---------------------|:---------------------:|:--------------------:|
    | Image node for Join |      PASS                 |       PASS               |
    | Existing join mode  |        PASS               |         PASS             |

   </details>

2. <details><summary>Negative case of ISO installation mode $~~$ $\color{green}{\textsf{PASS}}$</summary>

      - [ ] Reboot after installed binary mode
      - [ ] Shutdown / Boot after installed binary mode and test the join mode

   </details>

3. <details><summary>Streaming mode with manual boot installation $~~$ $\color{green}{\textsf{PASS}}$</summary>


    We can use `manual install` with boot option to use the streaming image to complete 
    
     - Install Harvester in create mode
     - Join Harvester to the existing node  

         ![image](https://github.com/harvester/harvester/assets/29251855/341052c1-4ce0-4ab9-9428-779e4874b00b)
    
         ![image](https://github.com/harvester/harvester/assets/29251855/e199f8f1-4087-4904-8339-dacdd53ba4ed)

   </details>

4. <details><summary>Streaming mode with import from storage installation $~~$ $\color{green}{\textsf{PASS}}$</summary>


    We can use `the pre-build streaming image file to install Harvester from import storage option 
    
     - Install Harvester in create mode
     - Join Harvester to the existing node  

        ![image](https://github.com/harvester/harvester/assets/29251855/1e866057-c27a-40ec-b7e5-f2d9705d6099)
        ![image](https://github.com/harvester/harvester/assets/29251855/49fc70fe-e957-4437-9512-705dfd183111)

   </details>


### Test Information
* Test Environment: Single and two nodes Harvester on local kvm machine
* Harvester version: `v1.2.0-rc4` and `v1.2-37aeb5ff`

### Verify Steps

  1. <details><summary>ISO install in image mode</summary>
        
        1. Launch vm and select local install media
        
            ![image](https://github.com/harvester/harvester/assets/29251855/ff839566-e3d4-4e9d-87a8-be3628817057)
 
        2. Select Install Harvester binary only
        
            ![image](https://github.com/harvester/harvester/assets/29251855/221c9933-8f6a-41dd-a2af-c8979dfaaaa8)
 
        3. provide disk information -> node password 
        4. Finish the image node installation 
        5. Installation will automatically boot up vm machine
        6. Select the `Create the Harvester cluster` 
        7. Finish the remaining steps of configuration 
        8. Check can install Harvester in create mode correctly
        9. Repeat step 1 ~ 5
        10. Select `Join an existing Harvester Cluster`
        11. Check can join the node to the previous created Harvester cluster
       

 </details> 

  2. <details><summary>Streaming mode: Install from manual boot option</summary>
        
        1. Prepare a pre-builded Harvester image file (e.g harvester-v1.2.0-rc4-amd64.raw.zst)   
       
        2. Download Harvester vmlinuz, initrd and squashfs from https://github.com/harvester/harvester/releases
        
        3. On the Harvester installation file path, launch python3 http server `python3 -m http.server`
           
        
        4. For create node, create the `userdata.yaml` file with the following content 
              ```
            #cloud-config
            scheme_version: 1
            token: token           # replace with a desired token
            os:
              password: p@ssword   # replace with a your password
              ntp_servers:
              - 0.suse.pool.ntp.org
              - 1.suse.pool.ntp.org
            install:
              mode: create
              automatic: true
              vip_mode: dhcp
            ```
 
        5. Launch VM and select `Manual install` 
        
            ![image](https://github.com/harvester/harvester/assets/29251855/6b06a409-8c7f-46af-a7c3-763330c6aecb)
            
        6. Change disk type to virtual disk
        
            ![image](https://github.com/harvester/harvester/assets/29251855/0c45a2fb-84ce-4ba5-b492-41d83867a9cd)

        7. Enable boot option, add vmlinuz and initrd file path 
        
            ![image](https://github.com/harvester/harvester/assets/29251855/4e960334-c952-45cd-bbff-4d3e53f17a37)

        8. Add the following Kernel argument 
        
        ```
          ip=dhcp net.ifnames=1 rd.cos.disable rd.noverifyssl root=live:http://192.168.122.1:8000/harvester-v1.2.0-rc4-rootfs-amd64.squashfs harvester.install.automatic=true harvester.scheme_version=1 harvester.install.device=/dev/vda  harvester.os.password=password harvester.install.raw_disk_image_path=http://192.168.122.1:8000/harvester-v1.2.0-rc4-amd64.raw.zst harvester.install.mode=install console=tty1 harvester.install.tty=tty1 harvester.install.config_url=http://192.168.122.1:8000/userdata.yaml harvester.install.management_interface.interfaces="name:ens3" harvester.install.management_interface.method=dhcp harvester.install.management_interface.bond_options.mode=balance-tlb harvester.install.management_interface.bond_options.miimon=100 harvester.install.power_off=true
        ```
        
        9. Boot up vm and wait the installation complete
        10. After complete, VM will automatically shutdown
        11. Edit VM setting, uncheck the `boot option`
            ![image](https://github.com/harvester/harvester/assets/29251855/5be8a717-d1c9-4595-9c16-443a25659659)

        12. Check can create Harvester node correctly
        
            ![image](https://github.com/harvester/harvester/assets/29251855/341052c1-4ce0-4ab9-9428-779e4874b00b)
            
            ![image](https://github.com/harvester/harvester/assets/29251855/e199f8f1-4087-4904-8339-dacdd53ba4ed)
        
        13. For join node scenario, you need to prepare another `userdata_join.yaml`
              ```
                #cloud-config
                scheme_version: 1
                server_url: https://192.168.122.173:443 # replace with the master node VIP
                token: token           # replace with a desired token
                os:
                  password: p@ssword   # replace with a your password
                  ntp_servers:
                  - 0.suse.pool.ntp.org
                  - 1.suse.pool.ntp.org
                  hostname: join-node
                install:
                  mode: join
                  automatic: true
                  vip_mode: dhcp
              ```
        14. And use this kernel args value in boot option instead 

        ```
          ip=dhcp net.ifnames=1 rd.cos.disable rd.noverifyssl root=live:http://192.168.122.1:8000/harvester-v1.2.0-rc4-rootfs-amd64.squashfs harvester.install.automatic=true harvester.scheme_version=1 harvester.install.device=/dev/vda  harvester.os.password=password harvester.install.raw_disk_image_path=http://192.168.122.1:8000/harvester-v1.2.0-rc4-amd64.raw.zst harvester.install.mode=install console=tty1 harvester.install.tty=tty1 harvester.install.config_url=http://192.168.122.1:8000/userdata_join.yaml harvester.install.management_interface.interfaces="name:ens3" harvester.install.management_interface.method=dhcp harvester.install.management_interface.bond_options.mode=balance-tlb harvester.install.management_interface.bond_options.miimon=100 harvester.install.power_off=true
        ```

        15. Then follow the same steps as create mode to join the Harvester cluster

            
            
        
          
       

 </details> 


  3. <details><summary>Streaming mode: Install from importing storage</summary>
        
        1. Prepare a pre-builded Harvester image file (e.g **harvester-v1.2.0-rc4-amd64.raw.zst**)   
       
        2. Prepare a `init.iso` and place in `/var/lib/libvirt/images` 
        
        3. Prepare a user-data in /tmp with content (for create mode)
        ```
            #cloud-config
            scheme_version: 1
            token: token # replace with a desired token
            os:
              password: p@ssword # replace with a your password
              ntp_servers:
                - 0.suse.pool.ntp.org
                - 1.suse.pool.ntp.org
            install:
              automatic: true
              mode: create
              vip_mode: dhcp
              management_interface:
                interfaces:
                  - name: ens3
                default_route: true
                method: dhcp
                bond_options:
                  mode: balance-tlb
                  miimon: 100
        ```
        
        4. For create node, extract the `harvester-v1.2.0-rc4-amd64.raw.zst` to raw image with command: 
             ```
              $ unzstd  harvester-v1.2.0-rc4-amd64.raw.zst
             ```
 
        5. Create an empty meta-data in /tmp
        
            `touch /tmp/meta-data` 
            
        6. Generate the cloud-init iso for create mode, named with `create.iso` 
            On Linux, you can use the command: 
            ```
              mkisofs -output /tmp/create.iso -volid cidata -joliet -rock /tmp/user-data /tmp/meta-data
            ```
            On Mac, you can use the command: 
            ```
            genisoimage  -output /tmp/create.iso  -volid cidata -joliet -rock /tmp/user-data /tmp/meta-data
            ```

        7. Copy the create.iso to your image path 
        
        8. Launch VM and select `Import existing disk`  
        
            ![image](https://github.com/harvester/harvester/assets/29251855/79a89b2a-12ca-46b9-b3f0-2b3a9dbf2738)

        9. Add the extracted raw image file
            ![image](https://github.com/harvester/harvester/assets/29251855/be9f93f5-0c7b-42da-a852-00b0fa9098eb)

        10. Change the disk type to `VirtIO`
        
            ![image](https://github.com/harvester/harvester/assets/29251855/f8383e4d-f858-4258-b409-f0f0275fd075)

        11. Add a new disk device, and select `CD ROM`, add `create.iso` into the path
        
            ![image](https://github.com/harvester/harvester/assets/29251855/6f586a3d-2735-4277-b41c-8fe5b37e5cf3)

        12. Launch the VM 
        13. Check can install Harvester correctly in create mode 
        14. For Join node, prepare a user-data in /tmp with content (for join mode) 
              (**you must use the same name user-data to replace the original one**)
              
              ```
              #cloud-config
                scheme_version: 1
                server_url: https://192.168.122.224:443 # Replace with the VIP 
                token: token # replace with a desired token
                os:
                  password: p@ssword # replace with a your password
                  ntp_servers:
                    - 0.suse.pool.ntp.org
                    - 1.suse.pool.ntp.org
                  hostname: join-node
                install:
                  automatic: true
                  mode: join
                  management_interface:
                    interfaces:
                      - name: ens3
                    default_route: true
                    method: dhcp
                    bond_options:
                      mode: balance-tlb
                      miimon: 100
              ```
              
        15. test 
        16. For create node, extract the `harvester-v1.2.0-rc4-amd64.raw.zst` to raw image with command: 
             ```
              $ unzstd  harvester-v1.2.0-rc4-amd64.raw.zst -o join-harvester-v1.2.0-rc4-amd64.raw
             ```
        17. Generate the cloud-init iso for join mode, named with `join.iso` 
            On Linux, you can use the command: 
            ```
              mkisofs -output /tmp/join.iso -volid cidata -joliet -rock /tmp/user-data /tmp/meta-data
            ```
            On Mac, you can use the command: 
            ```
            genisoimage  -output /tmp/join.iso  -volid cidata -joliet -rock /tmp/user-data /tmp/meta-data
            ```
        18.  Copy the create.iso to your image path 
        19.  Then follow the same steps as create mode to join the Harvester cluster
        20. Check can install Harvester correctly in create mode 

            ![image](https://github.com/harvester/harvester/assets/29251855/1e866057-c27a-40ec-b7e5-f2d9705d6099)
            
            ![image](https://github.com/harvester/harvester/assets/29251855/49fc70fe-e957-4437-9512-705dfd183111)



 </details> 

### Additional Context
I got the experience during testing image mode:
1. To use image mode in iso installation, suggest use disk size larger than 350Gi (400Gi in this case)
2. To use image mode in streaming, suggest use disk size larger then 350 Gi (The compressed zstd file is 18.3Gi, after extracted it will grow to 283Gi)

**Comment by Martin-Weiss** [comment]
Where can we find the documentation for this new feature (image based installation) and how can we automate the configuration of the machine in this case?
